kqueen/engines/heat.py

```diff
@@ -0,0 +1,178 @@
+from collections import namedtuple
+import importlib
+import yaml
+import logging
+import sys
+import os
+
+from kqueen.config import current_config
+import openstack
+
+from .base import BaseEngine
+
+logger = logging.getLogger('kqueen_api')
+config = current_config()
+
+
+STATE_MAP = {
+    'IN_PROGRESS': config.get('CLUSTER_PROVISIONING_STATE'),
+    'COMPLETE': config.get('CLUSTER_OK_STATE'),
+}
+
+
+class Heat:
+    templates = {
+        "default": "kqueen.engines.resources.heat.default",
+        "empty": "kqueen.engines.resources.heat.empty",
+    }
+
+    def __init__(self, **kwargs):
+        self.connection = openstack.connection.Connection(**kwargs)
+
+    def _get_template_and_files(self, name):
+        # NOTE(sskripnick)
+        # we need to custom load and decode all files because of this bug:
+        # https://storyboard.openstack.org/#!/story/2002002
+        package = importlib.import_module(self.templates[name])
+        path = list(package.__path__)[0]
+        files = {}
+        for root, dirs, _files in os.walk(path):
+            for f in _files:
+                fpath = os.path.join(root, f)
+                files["file://" + fpath] = open(fpath, "rb").read().decode("utf8")
+        return os.path.join(path, "template.yaml"), files
+
+    def create(self, *, name, template_name, ssh_key_name, server_image):
+        template_file, files = self._get_template_and_files(template_name)
+        try:
+            self.connection.create_stack(
+                name=name,
+                template_file=template_file,
+                files=files,
+                ssh_key_name=ssh_key_name,
+                server_image=server_image,
+            )
+        except openstack.exceptions.BadRequestException as e:
+            # NOTE(sskripnick)
+            # handle this exception separately because useful information
+            # may be found only in e.details
+            logger.exception("Failed to create stack: %s" % e.details)
+            return False, e.details
+        except Exception as e:
+            logger.exception("Failed to create stack:")
+            return False, e        
+    
+    def get(self, stack_name):
+        return self.connection.get_stack(stack_name)
+
+    def delete(self, name):
+        return True, None
+
+
+class OpenstackHeatEngine(BaseEngine):
+    name = "openstack_heat"
+    verbose_name = "Openstack Heat Engine"
+    parameter_schema = {
+        "cluster": {
+            "template": {
+                "type": "select",
+                "label": "Template",
+                "choices": (
+                    ("default", "Heat-Kubernetes"),
```

Comment:
imo naming is unclear

Reply:
I agree, that here we can allow user to upload a template file